### DISCUSSION TITLE
Installation auf Synology NAS

### DISCUSSION THREAD

**Discussion by klar574** [urlaubsverwaltung/zeiterfassung#610] [Q&A]
Hallo zusammen, 
hat es schon irgendjemand geschafft die Zeiterfassung auf einer NAS zu installieren? Ich habe es mit Portainer geschafft zu installieren, allerdings schmiert der Container Zeiterfassung ab. Es gibt hier wohl Probleme mit der Datenbankverbindung. 
Folgendes habe ich an der yml Datei angepasst:
01. postgres: (die installation mit Port 5432:5432 wurde abgebrochen, weil der Port belegt war)
    image: postgres:16.1
    ports:
      - '5433:5432'

02. keycloak: (auf der NAS wurden die Ordner erstellt)
volumes:
      - /volume1/docker/keycloak/import:/opt/keycloak/data/import

Fehlermeldungen vom Container:

date	stream	content
2024/03/06 13:31:34	stdout	Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'clientRegistrationRepository' defined in class path resource [org/springframework/boot/autoconfigure/security/oauth2/client/servlet/OAuth2ClientRegistrationRepositoryConfiguration.class]: Failed to instantiate [org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository]: Factory method 'clientRegistrationRepository' threw exception with message: Unable to resolve Configuration with the provided Issuer of "http://localhost:8090/realms/zeiterfassung-realm"
2024/03/06 13:31:34	stdout	Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'oidcClientInitiatedLogoutSuccessHandler' defined in class path resource [de/focusshift/zeiterfassung/security/SecurityBeanConfiguration.class]: Unsatisfied dependency expressed through method 'oidcClientInitiatedLogoutSuccessHandler' parameter 0: Error creating bean with name 'clientRegistrationRepository' defined in class path resource [org/springframework/boot/autoconfigure/security/oauth2/client/servlet/OAuth2ClientRegistrationRepositoryConfiguration.class]: Failed to instantiate [org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository]: Factory method 'clientRegistrationRepository' threw exception with message: Unable to resolve Configuration with the provided Issuer of "http://localhost:8090/realms/zeiterfassung-realm"
2024/03/06 13:31:34	stdout	org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'securityWebConfiguration' defined in file [/workspace/BOOT-INF/classes/de/focusshift/zeiterfassung/security/SecurityWebConfiguration.class]: Unsatisfied dependency expressed through constructor parameter 2: Error creating bean with name 'oidcClientInitiatedLogoutSuccessHandler' defined in class path resource [de/focusshift/zeiterfassung/security/SecurityBeanConfiguration.class]: Unsatisfied dependency expressed through method 'oidcClientInitiatedLogoutSuccessHandler' parameter 0: Error creating bean with name 'clientRegistrationRepository' defined in class path resource [org/springframework/boot/autoconfigure/security/oauth2/client/servlet/OAuth2ClientRegistrationRepositoryConfiguration.class]: Failed to instantiate [org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository]: Factory method 'clientRegistrationRepository' threw exception with message: Unable to resolve Configuration with the provided Issuer of "http://localhost:8090/realms/zeiterfassung-realm"
2024/03/06 13:31:34	stdout	2024-03-06T12:31:34.365Z ERROR 1 --- [           main] o.s.boot.SpringApplication               : Application run failed
2024/03/06 13:31:34	stdout	Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
2024/03/06 13:31:34	stdout	2024-03-06T12:31:34.155Z  WARN 1 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'securityWebConfiguration' defined in file [/workspace/BOOT-INF/classes/de/focusshift/zeiterfassung/security/SecurityWebConfiguration.class]: Unsatisfied dependency expressed through constructor parameter 2: Error creating bean with name 'oidcClientInitiatedLogoutSuccessHandler' defined in class path resource [de/focusshift/zeiterfassung/security/SecurityBeanConfiguration.class]: Unsatisfied dependency expressed through method 'oidcClientInitiatedLogoutSuccessHandler' parameter 0: Error creating bean with name 'clientRegistrationRepository' defined in class path resource [org/springframework/boot/autoconfigure/security/oauth2/client/servlet/OAuth2ClientRegistrationRepositoryConfiguration.class]: Failed to instantiate [org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository]: Factory method 'clientRegistrationRepository' threw exception with message: Unable to resolve Configuration with the provided Issuer of "http://localhost:8090/realms/zeiterfassung-realm"
2024/03/06 13:30:29	stderr	Picked up JAVA_TOOL_OPTIONS: -Djava.security.properties=/layers/paketo-buildpacks_bellsoft-liberica/java-security-properties/java-security.properties -XX:+ExitOnOutOfMemoryError -XX:ActiveProcessorCount=8 -XX:MaxDirectMemorySize=10M -Xmx34265138K -XX:MaxMetaspaceSize=129009K -XX:ReservedCodeCacheSize=240M -Xss1M -XX:+UnlockDiagnosticVMOptions -XX:NativeMemoryTracking=summary -XX:+PrintNMTStatistics -Dorg.springframework.cloud.bindings.boot.enable=true

Danke und Gruß Jens

**Comment by Scorpoon** [comment]
Ja ist leider sehr viel mit trial & error verbunden weil vieles einfach nicht beschrieben ist.

Den Fehler hatte ich auch, normalerweise müsste zumindest Keycloak laufen oder?
Geh auf die Keycloak Seite (die auf deiner NAS läuft), melde dich mit den Zugangsdaten an die du angegeben hast.

Dann gibt es links oben ein Dropdown, klick das an und erstelle einen neuen Realm. Als resource file kopierst du den Inhalt von diesem Github Repo im Ordner "docker/keycloack/export" (ich hab die "zeiterfassung-realm-realm-json" Datei verwendet.

Danach sollte sich denk ich auch der zeiterfassungs service starten lassen.

**Reply by klar574** [reply]
Hi Scorpoon,

Super, ja Danke das hat schon mal jetzt soweit geklappt. Jetzt erscheint allerdings beim Aufruf der Seite Ungültige Parameter: redirect_uri

SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_DEFAULT_REDIRECT-URI: http://192.168.xxx.xxx:8080/login/oauth2/code/{registrationId}

So sieht der Eintrag jetzt bei mir aus. Hab halt überall wo localhost eingetragen war, die IP meiner NAS genommen

**Reply by afarud** [reply]
Hallo zusammen,
kann einer vorn euch hier beschreiben, wie ihr es bisher geschafft habt? 
Also ich finde es echt nicht docker like, wenn es danach noch hundert Einstellungen geben muss. 
Ich habe etliche docker am laufen und bis auf diesen hier, komme ich relativ klar. 

Weder mit dem portainer, noch mit Synology will mir diese App gelingen. 
Danek euch vorab.

afarud

**Reply by Scorpoon** [reply]
Habs aufgegeben nachdem die Rechteeinstellungen anscheinend nicht richtig funktioniert haben oder was auch immer. Habs dann in die Tonne geworfen

**Reply by afarud** [reply]
Danke für die Antwort, so spare ich mir echt Lebenszeit. 
Vielen Dank. 
afarud

**Reply by klar574** [reply]
Vermutlich liegt das Ziel der Entwickler lieber am Verkauf des Produktes. Sehr schade. Weil da gibt Zick andere Anbieter

**Comment by Scorpoon** [comment]
🤣
